src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -779,7 +779,38 @@ pub trait SnapshotOps: StoreWrapper + 'static {
     let SubsetParams { globs } = params;
     snapshot_glob_match(self.clone(), digest, globs).await
   }
+
+  async fn create_empty_dir(&self, path: RelativePath) -> Result<Digest, SnapshotOpsError> {
+    let path: PathBuf = path.into();
+    let mut digest = EMPTY_DIGEST;
+    let mut components = path.components();
+    while let Some(parent) = components.next_back() {
+      let parent = match &parent {
+        Component::Normal(p) => p,
+        x => {
+          return Err(
+            format!(
+              "Cannot add component \"{:?}\" of directory path `{}`.",
```

Comment:
When would this happen? Is there more useful context we can provide?

Reply:
I'm not sure - honestly, I copied it from the `add_prefix` code.

Reply:
That's fine! If we haven't seen it come up despite being used there already I think it might be ok.

Reply:
It can't happen since path is a RelativePath which ensures only Normal components survive:
https://github.com/pantsbuild/pants/blob/7c8947d8fe46c4d8ab69e3229bcab1ff89b03107/src/rust/engine/fs/src/lib.rs#L101-L125

Maybe you could leverage that guaranty and just use path.iter() to drive the loop over known-good OsStr values directly. This is fine since the Err can't happen, but it is a bit noisy.

Reply:
Much better! This allowed some simplifications elsewhere too. Thanks!